ckanext/validation/tests/test_uploader.py

```diff
@@ -0,0 +1,347 @@
+        extra_environ=_get_env(),
+        data=data,
+    )
+
+    resource = call_action("resource_show", id=dataset["resources"][0]["id"])
+
+    assert resource["schema"] == {
+        "fields": [
+            {"name": "e", "type": "integer"},
+            {"name": "f", "type": "integer"},
+        ]
+    }
+
+
+def test_update_removes_schema(app):
+
+    dataset = Dataset()
+
+    data = {
+        "upload": (io.BytesIO(bytes(VALID_CSV.encode("utf8"))), "valid.csv"),
+    }
+
+    app.post(
+        url=_new_resource_upload_url(dataset["id"]), extra_environ=_get_env(), data=data
+    )
+
+    dataset = call_action("package_show", id=dataset["id"])
+
+    assert dataset["resources"][0]["schema"]
+
+    data = {
+        "upload": (io.BytesIO(b"test file"), "some.txt"),
+    }
+
+    app.post(
+        url=_edit_resource_upload_url(dataset["id"], dataset["resources"][0]["id"]),
+        extra_environ=_get_env(),
+        data=data,
+    )
+
+    resource = call_action("resource_show", id=dataset["resources"][0]["id"])
+
+    assert "schema" not in resource
+
+
+def test_update_upload_no_tabular_no_schema(app):
+
+    resource = Resource()
+
+    data = {
+        "upload": (io.BytesIO(b"test file"), "some.txt"),
+    }
+
+    app.post(
+        url=_edit_resource_upload_url(resource["package_id"], resource["id"]),
+        extra_environ=_get_env(),
+        data=data,
+    )
+
+    resource = call_action("resource_show", id=resource["id"])
+
+    assert resource["format"] == "TXT"
+    assert resource["url_type"] == "upload"
+
+    assert "schema" not in resource
+
+
+@responses.activate
+def test_update_url_with_schema(app):
+    url = "https://example.com/valid.csv"
+
+    responses.add(responses.GET, url, body=VALID_CSV)
+    responses.add_passthru(config["solr_url"])
+
+    resource = Resource()
+
+    data = {
+        "url": url,
+        # CKAN does not refresh the format, see https://github.com/ckan/ckan/issues/7415
+        "format": "CSV",
+    }
+
+    app.post(
+        url=_edit_resource_upload_url(resource["package_id"], resource["id"]),
+        extra_environ=_get_env(),
+        data=data,
+    )
+
+    resource = call_action("resource_show", id=resource["id"])
+
+    assert resource["format"] == "CSV"
+    assert resource["url_type"] is None
+
+    assert resource["schema"] == {
+        "fields": [
+            {"name": "a", "type": "integer"},
+            {"name": "b", "type": "integer"},
+            {"name": "c", "type": "integer"},
+            {"name": "d", "type": "integer"},
+        ]
+    }
+
+
+@responses.activate
+def test_update_url_no_tabular_no_schema(app):
+    url = "https://example.com/some.txt"
+
+    responses.add(responses.GET, url, body="some text")
+    responses.add_passthru(config["solr_url"])
+
+    resource = Resource()
+
+    data = {
+        "url": url,
+        # CKAN does not refresh the format, see https://github.com/ckan/ckan/issues/7415
```

Comment:
with the new widget, I'm retrieving the format when the user replace the uploaded resource and changing it on the resource edit form. if the user save the changes the new format is updated but I agree with you, it would be better to do it in the backend